site-src/geps/gep-1762.md

```diff
@@ -0,0 +1,309 @@
+This is typically not a concern.
+However, if an implementation is aware of specific meanings of certain labels or annotations, they MAY
+exclude these from irrelevant resources.
+
+#### Arbitrary Customization
+
+Currently, to provide arbitrary customizations, Gateway API provides a few mechanisms:
+
+* GatewayClass has a `parametersRef` field, allowing to point to an arbitrary other resource.
+
+  This *can* be used to make arbitrary customizations, but comes with a number of issues:
+  * GatewayClass is cluster scoped. This means that any user configuring a namespace-scoped Gateway
+    may need to make a new cluster-scoped resources, which makes RBAC extremely challenging.
+    An analogy would be to have a `DeploymentClass` to configure CPU requests, instead of a field in each `Deployment`.
+  * Many levels of indirection. To configure my `Gateway`, I need a `Gateway`, `GatewayClass`, and `SomeCustomParamsType`.
+  * This is intended to be immutable according to the `GatewayClass` spec. This makes it challenging to operate and implement.
+  * There is no common type to reference; each implementation would make their own.
+
+* Annotations are always an option, but one that is counter to many Gateway API goals.
+
+* Policy attachment could be used to modify a `Gateway`, but this is a fairly unintended usage of policy attachment.
+
+* Modifying and merging user values and controller values.
+  Users can manually create the generated resources and have them merged in with the controller values.
+
+  With careful use of Server Side Apply, this could be done in any order. However, this comes with downsides:
+  * No good way to have implementation provide "defaults" and "overrides".
+  * Order may matter; if implementation is first we may temporarily have unsafe settings (for example, run as root later overwritten to run as non-root).
+  * This type of ordering dependency is tricky operationally, especially with GitOps.
+
+None of these mechanisms provide a satisfying solution.
+
+Some options:
+
+* Add a `infrastructure` field into `Gateway`. This would have a few common configuration settings in the top level, then delegate to an arbitrary type:
+
+  ```yaml
+  spec:
+    infrastructure:
+      # Define the scope of the Gateway (GEP-1651)
+      scope: InCluster
+      # Define the version to use. The meanings of version is implementation specific
+      version: v1.2
+      # Number of replicas for the Gateway. The meaning is implementation specific; for Deployments this SHOULD be `.spec.replicas`.
+      replicas: 2
+      parametersRef:
+        kind: SomeCustomType
+        name: my-params
+    listeners:
+    - ...
+  ```
+
+### Resource Attachment
+
+Resources Generated in response to the `Gateway` are guaranteed to have two attributes:
+* A `gateway.networking.k8s.io/metadata.name: <NAME>` label.
+* A name `<NAME>-<GATEWAY CLASS>`.
+
+The generated resources MUST be in the same namespaces as the `Gateway`.
```

Comment:
We should probably discuss this, may be too expensive and not clear if it is necessary. 

If we allow external multi-tenant - we can probably support per-cluster gateway too. Implementations should be free to generate the deployment in a central namespace if they can preserve the isolation semantics.

Reply:
I'm very reluctant to allow ways to _explicitly_ request crossing namespace boundaries in the absence of a two-way handshake (which is not practical here). Maybe a better way to write this is "_If_ Kubernetes resources are generated to satisfy a Gateway request, they MUST be in the same namespace as the Gateway".

Cilium's Gateway API support doesn't generate a Deployment to provision an Envoy instance, but does provision CiliumEnvoyConfig resources to control the per-node Envoy instances, and _those_ are created in the same namespace as the Gateway.